### PR TITLE
Add ActiveMQ classic monitoring

### PR DESCRIPTION
**Pull request by CzyerChen**
### <Feature description>https://github.com/apache/skywalking/discussions/12063
- [x] Update the documentation to include this new feature.
- [x] Tests(including UT, IT, E2E) are added to verify the new feature. 
ActiveMQ single-node/ 2 single-nodes/ shared-file-master-slave have been tested.
- [x] If it's UI related, attach the screenshots below.
- [x] Update the [`CHANGES` log](https://github.com/apache/skywalking/blob/master/docs/en/changes/changes.md).

For ActiveMQ Classic cluster:

image

- System Load: range in [0, 100].
- Thread Count: the number of threads currently used by the JVM.
- Heap Memory: capacity of heap memory.
- GC: memory of ActiveMQ is managed by Java's garbage collection (GC) process. Compatible with JDK 1.6-17 with ActiveMQ classic(5.10.x ~ 6.1.0)
- Enqueue/Dequeue/Dispatch/Expired Rate: growth rate of messages in different states.
- Average/Max Enqueue Time: time taken to join the queue.

For ActiveMQ Classic broker:

image

- Uptime: duration of the node.
- State: 1 = slave node, 0 = master node.
- Current Connentions: number of connections.
- Current Producer/Consumer Count: number of current producers/consumers.
- Increased Producer/Consumer Count: number of increased producers/consumers.
- Enqueue/Dequeue Count: number of enqueue and dequeue.
- Enqueue/Dequeue Rate: rate of enqueue and dequeue.
- Memory Percent Usage: amount of memory space used by undelivered messages.
- Store Percent Usage: space used by pending persistent messages.
- Temp Percent Usage: space used by non-persistent messages.
- Average/Max Message Size: number of messages.
- Queue Size: number of messages in the queue.

For ActiveMQ Classic destination:

image

- Produser/Consumer Count: number of producers/Consumers.
- Queue Size: unacknowledged messages of the queue.
- Memory usage: usage of memory.
- Enqueue/Dequeue/Dispatch/Expired/Inflight Count: number of messages in different states.
- Average/Max Message Size: number of messages.
- Average/Max Enqueue Time: time taken to join the queue.

Due to the shortcomings of the activemq-5.x agent plugin, it can not suit hierarchy definition now.
I have to fix the agent part later, then push the hierarchy definition.

### CODE REVIEW

File: oap-server/server-starter/src/main/resources/ui-initialized-templates/activemq/activemq-destination.json [hunk 1]
```diff
@@ -0,0 +1,286 @@
...
+          "metricTypes": [],
+          "metrics": [],
+          "metricMode": "Expression",
+          "moved": false,
+          "expressions": [
+            "meter_activemq_destination_producer_count"
+          ],
+          "typesOfMQE": [
+            "TIME_SERIES_VALUES"
+          ],
```
Comment (wankai123) [inline oap-server/server-starter/src/main/resources/ui-initialized-templates/activemq/activemq-destination.json:23]: metricMode, metrics, metricTypes, typesOfMQE... have been removed from the latest UI, please sync UI and re-export template.

Comment (CzyerChen) [reply oap-server/server-starter/src/main/resources/ui-initialized-templates/activemq/activemq-destination.json:23]: ok

**Comment by wu-sheng** [comment]
The UI submodule seems to be changed unexpectedly?

File: oap-server/server-starter/src/main/resources/ui-initialized-templates/activemq/activemq-cluster.json [hunk 2]
```diff
@@ -0,0 +1,697 @@
...
+                  ],
+                  "graph": {
+                    "type": "Area",
+                    "opacity": 0.4,
+                    "showXAxis": true,
+                    "showYAxis": true
+                  },
+                  "widget": {
+                    "name": "GcOldCollectionTime",
+                    "title": "Gc Old Collection Time",
```
Comment (wankai123) [inline oap-server/server-starter/src/main/resources/ui-initialized-templates/activemq/activemq-cluster.json:499]: suggest adding the units on the title.

**Comment by wu-sheng** [comment]
Another question is, what is the status of service hierarchy about this new active MQ monitoring?

File: oap-server/server-starter/src/main/resources/ui-initialized-templates/activemq/activemq-cluster.json (hunk 2 above)
Comment (wu-sheng) [reply oap-server/server-starter/src/main/resources/ui-initialized-templates/activemq/activemq-cluster.json:499]: `Gc` actually should be `GC`

**Comment by CzyerChen** [comment]
> Another question is, what is the status of service hierarchy about this new active MQ monitoring?

now:
virtual_mq service name: 192.168.208.2:61616
agent peer:  192.168.208.2:61616
dashboard service name: activemq.skywaling-showcase.svc.local

Currently, it is not compatible with building service hierarchy by using servicename.

**Comment by wu-sheng** [comment]
That rule should base on k8s deployment. Not VM.

File: oap-server/server-starter/src/main/resources/ui-initialized-templates/activemq/activemq-destination.json (hunk 1 above)
Comment (CzyerChen) [reply oap-server/server-starter/src/main/resources/ui-initialized-templates/activemq/activemq-destination.json:23]: Any changes about `table`

es monitoring panel about node health, not shown as expected

image

Comment (CzyerChen) [reply oap-server/server-starter/src/main/resources/ui-initialized-templates/activemq/activemq-destination.json:23]: image

Comment (wu-sheng) [reply oap-server/server-starter/src/main/resources/ui-initialized-templates/activemq/activemq-destination.json]: @Fine0830 Could you recheck this?

**Comment by CzyerChen** [comment]
> The UI submodule seems to be changed unexpectedly?

Reverted.

File: oap-server/server-starter/src/main/resources/ui-initialized-templates/activemq/activemq-destination.json (hunk 1 above)
Comment (wu-sheng) [reply oap-server/server-starter/src/main/resources/ui-initialized-templates/activemq/activemq-destination.json]: Fixed by https://github.com/apache/skywalking-booster-ui/pull/389

File: test/e2e-v2/cases/activemq/docker-compose.yml [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+
+  amq:
+    image: apache/activemq-classic:6.0.1
+    hostname: amq
+    ports:
+      - "61616:61616"
+      - "8161:8161"
+    expose:
+      - 1616
+    volumes:
```
Comment (weixiang1862) [inline test/e2e-v2/cases/activemq/docker-compose.yml:36]: Seems JMX export of activemq is not out-of-the-box and needs some configurations, could you add this part to `setup/backend/backend-activemq-monitoring.md` doc or add a ref of this case from there.

**Comment by CzyerChen** [comment]
> That rule should base on k8s deployment. Not VM.

ActiveMQ agent side set  Tags.MQ_BROKER = raw IP address:port, rather than hostname:port

in K8S:
ACTIVEMQ-SERVER(ip:port) -> VIRTUAL_MQ( ip:port ) : support service hierarchy
VIRTUAL_MQ( ip:port ) -> ACTIVEMQ cluster monitoring(activemq.skywalking-showcase) : can not support service hierarchy
ACTIVEMQ cluster monitoring(activemq.skywalking-showcase) -> K8S_SERVICE(skywalking-showcase::activemq.skywalking-showcase): support service hierarchy

**Comment by wu-sheng** [comment]
Do we have IP-service name resolver to support in hierarchy detection? I think that would be useful.

**Comment by wu-sheng** [comment]
> > That rule should base on k8s deployment. Not VM.
> 
> ActiveMQ agent side set Tags.MQ_BROKER = raw IP address:port, rather than hostname:port
> 
> in K8S: ACTIVEMQ-SERVER(ip:port) -> VIRTUAL_MQ( ip:port ) : support service hierarchy VIRTUAL_MQ( ip:port ) -> ACTIVEMQ cluster monitoring(activemq.skywalking-showcase) : can not support service hierarchy ACTIVEMQ cluster monitoring(activemq.skywalking-showcase) -> K8S_SERVICE(skywalking-showcase::activemq.skywalking-showcase): support service hierarchy

So, the name of the ActiveMQ service is hard coded only?

**Comment by wankai123** [comment]
@CzyerChen  
> ACTIVEMQ-SERVER(ip:port) -> VIRTUAL_MQ( ip:port ) 

Where is the `ACTIVEMQ-SERVER(ip:port)` data from？

**Comment by CzyerChen** [comment]
> > > That rule should base on k8s deployment. Not VM.
> > 
> > 
> > ActiveMQ agent side set Tags.MQ_BROKER = raw IP address:port, rather than hostname:port
> > in K8S: ACTIVEMQ-SERVER(ip:port) -> VIRTUAL_MQ( ip:port ) : support service hierarchy VIRTUAL_MQ( ip:port ) -> ACTIVEMQ cluster monitoring(activemq.skywalking-showcase) : can not support service hierarchy ACTIVEMQ cluster monitoring(activemq.skywalking-showcase) -> K8S_SERVICE(skywalking-showcase::activemq.skywalking-showcase): support service hierarchy
> 
> About this, in k8s,
> 
> * VIRTUAL_MQ( ip:port ) could be service name, is the activeMQ plugin(Java agent) resolving the Domain name and gets the IP?
> * `ACTIVEMQ cluster monitoring(activemq.skywalking-showcase)` should be k8s service name, right?

yes

**Comment by CzyerChen** [comment]
> @CzyerChen
> 
> > ACTIVEMQ-SERVER(ip:port) -> VIRTUAL_MQ( ip:port )
> 
> Where is the `ACTIVEMQ-SERVER(ip:port)` data from？

Data reported by activeMQ plugin(Java agent).

**Comment by wu-sheng** [comment]
> > > > That rule should base on k8s deployment. Not VM.
> > > 
> > > 
> > > ActiveMQ agent side set Tags.MQ_BROKER = raw IP address:port, rather than hostname:port
> > > in K8S: ACTIVEMQ-SERVER(ip:port) -> VIRTUAL_MQ( ip:port ) : support service hierarchy VIRTUAL_MQ( ip:port ) -> ACTIVEMQ cluster monitoring(activemq.skywalking-showcase) : can not support service hierarchy ACTIVEMQ cluster monitoring(activemq.skywalking-showcase) -> K8S_SERVICE(skywalking-showcase::activemq.skywalking-showcase): support service hierarchy
> > 
> > 
> > About this, in k8s,
> > 
> > * VIRTUAL_MQ( ip:port ) could be service name, is the activeMQ plugin(Java agent) resolving the Domain name and gets the IP?
> > * `ACTIVEMQ cluster monitoring(activemq.skywalking-showcase)` should be k8s service name, right?
> 
> yes

Let's separate this in two ways. 
1. Service hierarchy is still working if you are building ActiveMQ server monitoring with K8s service monitoring. The hierarchy between those two isn't affected by the agent use case.
2. About the agent is reporting the service IP, is this a good practice? I am thinking this because usually virtual MQ should be a logic concept, if the MQ is restarted by k8s, the IP could change, but DNS name would not. Nowadays, k8s deployment is very popular, and it is the key use case of the hierarchy feature, we only need to focus on that. How about add `remote.ip` tag for the actual IP, and use `peer` for domain name?

**Comment by CzyerChen** [comment]
> ACTIVEMQ-SERVER

~~ACTIVEMQ-SERVER~~ should be `ACTIVEMQ-CLIENT` collected by agent side.

File: oap-server/server-starter/src/main/resources/hierarchy-definition.yml [hunk 4]
```diff
@@ -60,6 +60,9 @@ hierarchy:
   PULSAR:
     K8S_SERVICE: short-name
 
+  ACTIVEMQ:
+    K8S_SERVICE: short-name
```
Comment (wu-sheng) [inline oap-server/server-starter/src/main/resources/hierarchy-definition.yml:64]: How about the virtual MQ with this? Is that proposal possible?

**Comment by CzyerChen** [comment]
> > > > > That rule should base on k8s deployment. Not VM.
> > > > 
> > > > 
> > > > ActiveMQ agent side set Tags.MQ_BROKER = raw IP address:port, rather than hostname:port
> > > > in K8S: ACTIVEMQ-SERVER(ip:port) -> VIRTUAL_MQ( ip:port ) : support service hierarchy VIRTUAL_MQ( ip:port ) -> ACTIVEMQ cluster monitoring(activemq.skywalking-showcase) : can not support service hierarchy ACTIVEMQ cluster monitoring(activemq.skywalking-showcase) -> K8S_SERVICE(skywalking-showcase::activemq.skywalking-showcase): support service hierarchy
> > > 
> > > 
> > > About this, in k8s,
> > > 
> > > * VIRTUAL_MQ( ip:port ) could be service name, is the activeMQ plugin(Java agent) resolving the Domain name and gets the IP?
> > > * `ACTIVEMQ cluster monitoring(activemq.skywalking-showcase)` should be k8s service name, right?
> > 
> > 
> > yes
> 
> Let's separate this in two ways.
> 
> 1. Service hierarchy is still working if you are building ActiveMQ server monitoring with K8s service monitoring. The hierarchy between those two isn't affected by the agent use case.
> 2. About the agent is reporting the service IP, is this a good practice? I am thinking this because usually virtual MQ should be a logic concept, if the MQ is restarted by k8s, the IP could change, but DNS name would not. Nowadays, k8s deployment is very popular, and it is the key use case of the hierarchy feature, we only need to focus on that. How about add `remote.ip` tag for the actual IP, and use `peer` for domain name?

Now part1 supported:

image

part2 agent side needs to upgrade

image

image

image

**Comment by wu-sheng** [comment]
> part2 agent side needs to upgrade

I think we could put the rules here once for all? As currently, the agent side has a totally different release period. The next(9.3) agent release could be 3 months later.

**Comment by CzyerChen** [comment]
> > part2 agent side needs to upgrade
> 
> I think we could put the rules here once for all? As currently, the agent side has a totally different release period. The next(9.3) agent release could be 3 months later.

`  - VIRTUAL_MQ -> ACTIVEMQ` may have to check `cluster` mode, rather than single node. if client connected by `failover:(tcp://amq1:61616,tcp://amq2:61617)`, it may hard to continue service hierarchy.

**Comment by wu-sheng** [comment]
> ` - VIRTUAL_MQ -> ACTIVEMQ` may have to check `cluster` mode, rather than single node. if client connected by `failover:(tcp://amq1:61616,tcp://amq2:61617)`, it may hard to continue service hierarchy.

How could this be in k8s? I think K8s should be DNS oriented right? Even with a cluster, client should use service address like `activemq-svr.namespace.cluster`.

VM deployment is complex, we are hard to make service hierarchy working in that deployment out-of-box.

File: docs/en/swip/readme.md [hunk 5]
```diff
@@ -72,6 +72,7 @@ Next SWIP Number: 6
 
 ### Accepted SWIPs
 
+- [SWIP-6 Support ActiveMQ Monitoring](SWIP-6.md)
```
Comment (wu-sheng) [inline docs/en/swip/readme.md:75]: Could you update `Next SWIP Number: 6` to `Next SWIP Number: 8`? 7 was declared by https://github.com/apache/skywalking/discussions/12071, although it is not done yet.

Comment (CzyerChen) [reply docs/en/swip/readme.md:75]: OK

**Comment by wu-sheng** [comment]
@CzyerChen How about the latest question?

**Comment by CzyerChen** [comment]
> > ` - VIRTUAL_MQ -> ACTIVEMQ` may have to check `cluster` mode, rather than single node. if client connected by `failover:(tcp://amq1:61616,tcp://amq2:61617)`, it may hard to continue service hierarchy.
> 
> How could this be in k8s? I think K8s should be DNS oriented right? Even with a cluster, client should use service address like `activemq-svr.namespace.cluster`.
> 
> VM deployment is complex, we are hard to make service hierarchy working in that deployment out-of-box.

Normally we use failover transport to connect server node on client side, like `failover:(tcp://broker1.namespace.cluster:61616,tcp://broker2.namespace.cluster:61616,tcp://broker3.namespace.cluster:61616)?transportOptions` in K8S.

In Master-Slave or Broker cluster mode, If just configure one node of Broker on client side, the failover will not work automatically.

**Comment by wu-sheng** [comment]
Are you saying this is recommended deployment for ActiveMQ on k8s?
No unified domain namr for its cloud native deployment by adopting k8s DNS resolving mechanism?

**Comment by wu-sheng** [comment]
Typically in k8s, two or more deployment could be as one service, and failover could be done internally and natively.

**Comment by CzyerChen** [comment]
> Are you saying this is recommended deployment for ActiveMQ on k8s? No unified domain namr for its cloud native deployment by adopting k8s DNS resolving mechanism?

It needs to list the addresses of all brokers in the cluster now in ActiveMQ Classic(no specific description for K8S deployment). But after introducing `JGroups` auto-discovery from Artemis, the client no longer needs to list the addresses of all brokers in the cluster, only need the logical address of the cluster. 

For the Classic, there is no `nameserver` to provide a unified connection portal.

**Comment by wu-sheng** [comment]
OK, @wankai123 Let us skip that.

File: docs/en/setup/backend/backend-activemq-monitoring.md [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+2. The [jmx prometheus exporter](https://github.com/prometheus/jmx_exporter) collects metrics data from ActiveMQ classic, this exporter is intended to be run as a Java Agent, exposing a HTTP server and serving metrics of the local JVM.  
+3. OpenTelemetry Collector fetches metrics from jmx prometheus exporter via Prometheus Receiver and pushes metrics to
+   SkyWalking OAP Server via OpenTelemetry gRPC exporter.
+4. The SkyWalking OAP Server parses the expression with [MAL](../../concepts-and-designs/mal.md) to
+   filter/calculate/aggregate and store the results.
+
+## Setup
+
+1. [Enable JMX](https://activemq.apache.org/components/classic/documentation/jmx) in `activemq.xml`, the JMX remote port defaults to `1616`, you can change it through `ACTIVEMQ_SUNJMX_START`.
+2. Setup [jmx prometheus exporter](https://github.com/prometheus/jmx_exporter) which runs as a Java Agent(recommended) of ActiveMQ classic.If you work with docker, [refer here](https://github.com/bitnami/containers/tree/main/bitnami/jmx-exporter) for docker deployment.
```
Comment (wu-sheng) [inline docs/en/setup/backend/backend-activemq-monitoring.md:19]: Could you follow this https://github.com/apache/skywalking/pull/12109#discussion_r1566835326 and add more words here? And you could add a link to the e2e configuration as a reference.

Comment (CzyerChen) [reply docs/en/setup/backend/backend-activemq-monitoring.md:19]: OK

File: test/e2e-v2/cases/activemq/docker-compose.yml (hunk 3 above)
Comment (CzyerChen) [reply test/e2e-v2/cases/activemq/docker-compose.yml:36]: https://github.com/apache/skywalking/pull/12109/commits/3223307205f90fb00bb28e496eb56b0fb5b367be

updated.

File: oap-server/server-starter/src/main/resources/ui-initialized-templates/activemq/activemq-broker.json [hunk 7]
```diff
@@ -0,0 +1,355 @@
...
+            "children": [
+                {
+                    "x": 6,
+                    "y": 0,
+                    "w": 6,
+                    "h": 9,
+                    "i": "0",
+                    "type": "Widget",
+                    "expressions": [
+                        "meter_activemq_broker_state"
```
Comment (wankai123) [inline oap-server/server-starter/src/main/resources/ui-initialized-templates/activemq/activemq-broker.json:14]: please use the `latest or avg`... function to aggregate values if the widget doesn't show the time series values.

**Comment by wankai123** [comment]
The others are LGTM.

File: oap-server/server-starter/src/main/resources/ui-initialized-templates/activemq/activemq-broker.json (hunk 7 above)
Comment (CzyerChen) [reply oap-server/server-starter/src/main/resources/ui-initialized-templates/activemq/activemq-broker.json]: https://github.com/apache/skywalking/pull/12109/commits/67673bd1c40d106a05acd341e4414a27a5d5d637
updated

**Review by wu-sheng — APPROVED** [review]
lgtm

**Review by wankai123 — APPROVED** [review]

**Review by weixiang1862 — APPROVED** [review]

**Comment by wankai123** [comment]
@CzyerChen, Hi, I found some expressions didi not add `service_instance_id` in `activemq-cluster.yaml`, is that expected?

**Comment by CzyerChen** [comment]
> @CzyerChen, Hi, I found some expressions didi not add `service_instance_id` in `activemq-cluster.yaml`, is that expected?

In Master-Slave mode, cluster metrics only report from Master node, but JVM metrics will report from all nodes. So JVM metrics add `service_instance_id`, cluster metrics not.

**Pull request merged**